perf/rdm_bw.c

```diff
@@ -0,0 +1,471 @@
+
+err5:
+	free_ep_res();
+err4:
+	fi_close(&ep->fid);
+err3:
+	fi_close(&dom->fid);
+err2:
+	fi_close(&fab->fid);
+err1:
+	return ret;
+}
+
+static int init_av(void)
+{
+	void *addr;
+	size_t addrlen = 0;
+	int ret;
+
+	fi_getname(&ep->fid, NULL, &addrlen);
+	addr = malloc(addrlen);
+	assert(addr);
+	ret = fi_getname(&ep->fid, addr, &addrlen);
+	if (ret != 0) {
+		FT_PRINTERR("fi_getname", ret);
+		return ret;
+	}
+
+	addrs = malloc(numprocs * addrlen);
+	assert(addrs);
+
+	FT_Allgather(addr, addrlen, addrs);
+
+	fi_addrs = malloc(numprocs * sizeof(fi_addr_t));
+	assert(fi_addrs);
+
+	/* Insert address to the AV and get the fabric address back */
+	ret = fi_av_insert(av, addrs, numprocs, fi_addrs, 0, &fi_ctx_av);
+	if (ret != numprocs) {
+		FT_PRINTERR("fi_av_insert", ret);
+		return ret;
+	}
+
+	free(addr);
+
+	return 0;
+}
+
+int main(int argc, char *argv[])
+{
+	int i, j, peer;
+	int size, align_size;
+	char *s_buf, *r_buf;
+	double t_start = 0.0, t_end = 0.0, t = 0.0;
+	int op, ret;
+	buf_desc_t lbuf_desc;
+	ssize_t fi_rc;
+
+	FT_Init(&argc, &argv);
+	FT_Rank(&myid);
+	FT_Job_size(&numprocs);
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return -1;
+
+	while ((op = getopt(argc, argv, "f:h")) != -1) {
+		switch (op) {			
+		case 'f':
+			hints->fabric_attr->prov_name = strdup(optarg);
+			break;
+		case '?':
+		case 'h':
+			print_usage();
+			return -1;
+		}
+	}
+
+	hints->ep_attr->type	= FI_EP_RDM;
+	hints->caps		= FI_MSG | FI_RMA;
+	hints->mode		= ~0;
+	hints->domain_attr->mr_mode = FI_MR_BASIC;
+
+	if(numprocs != 2) {
+		if(myid == 0) {
+			fprintf(stderr, "This test requires exactly two processes\n");
+		}
+		FT_Finalize();
+		return -1;
+	}
+
+	/* Fabric initialization */
+	ret = init_fabric();
+	if(ret) {
+		fprintf(stderr, "Problem in fabric initialization\n");
+		return ret;
+	}
+
+	ret = init_av();
+	if(ret) {
+		fprintf(stderr, "Problem in AV initialization\n");
+		return ret;
+	}
+
+	/* Data initialization */
+	align_size = getpagesize();
+	assert(align_size <= MAX_ALIGNMENT);
+
+	s_buf = (char *) (((unsigned long) s_buf_original + (align_size - 1)) /
+				align_size * align_size);
+	r_buf = (char *) (((unsigned long) r_buf_original + (align_size - 1)) /
+				align_size * align_size);
+
+	ret = fi_mr_reg(dom, r_buf, MYBUFSIZE, FI_REMOTE_WRITE, 0, 0, 0, &r_mr, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_mr_reg", ret);
+		return -1;
+	}
+
+	lbuf_desc.addr = (uint64_t)r_buf;
+	lbuf_desc.key = fi_mr_key(r_mr);
+
+	rbuf_descs = (buf_desc_t *)malloc(numprocs * sizeof(buf_desc_t));
+
+	/* Distribute memory keys */
+	FT_Allgather(&lbuf_desc, sizeof(lbuf_desc), rbuf_descs);
+
+	ret = fi_mr_reg(dom, s_buf, MYBUFSIZE, FI_WRITE, 0, 0, 0, &l_mr, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_mr_reg", ret);
+		return -1;
+	}
+
+	if (myid == 0) {
+		fprintf(stdout, HEADER);
+		fprintf(stdout, "%-*s%*s\n", 10, "# Size", FIELD_WIDTH,
+				"Bandwidth (MB/s)");
+		fflush(stdout);
+	}
+
+	/* Bandwidth test */
+	for (size = 1; size <= MAX_MSG_SIZE; size *= 2) {
```

Comment:
Personally, I'm not a fan of testing that does sizes 1, 2, 4, 8, 16, 32, 64, etc.  It ends up putting too much emphasis on very small buffer sizes.  Yes, this is just my personal preference, but who really cares about the difference between 1 byte versus 2 versus 4 versus 8?  The other fabtests use a message size array that includes tests 2^x and 2^x + 2^(x-1).  Can you please look at reusing that size array?